service/authentication/src/main/java/org/eclipse/kapua/service/authentication/AuthenticationServiceBackEndCall.java

```diff
@@ -57,43 +63,84 @@ public class AuthenticationServiceBackEndCall {
-            logger.info("Login for user: {} - clientId: {}", authRequest.getUsername(), authRequest.getClientId());
+            logger.info("Login for clientId {} - user: {} - password: {} - client certificates: {}",
+                    authRequest.getClientId(),
+                    authRequest.getUsername(),
+                    Strings.isNullOrEmpty(authRequest.getPassword()) ? "yes" : "no",
+                    authRequest.getCertificates() != null ? "yes" : "no");
+
             ThreadContext.unbindSubject();
-            UsernamePasswordCredentials credentials = credentialFactory.newUsernamePasswordCredentials(
-                authRequest.getUsername(), 
-                authRequest.getPassword());
-            AccessToken accessToken = authenticationService.login(credentials);
-            Account account = KapuaSecurityUtils.doPrivileged(()-> accountService.find(accessToken.getScopeId()));
+
+            // User retrieval
+            User user = KapuaSecurityUtils.doPrivileged(() -> userService.findByName(authRequest.getUsername()));
+            if (user == null) {
+                throw new KapuaEntityNotFoundException(User.TYPE, authRequest.getUsername());
+            }
+
+            // Device connection retrieval
+            DeviceConnection deviceConnection = KapuaSecurityUtils.doPrivileged(() -> deviceConnectionService.findByClientId(user.getScopeId(), authRequest.getClientId()));
+
+            // Authentication type selection
+            String deviceConnectionAuthType;
+            if (deviceConnection != null) {
+                // From the DeviceConnection
+                deviceConnectionAuthType = deviceConnection.getAuthenticationType();
+            } else {
+                // Or the default from DeviceConnectionService ServiceConfiguration
+                Map<String, Object> deviceConnectionServiceConfigValues = KapuaSecurityUtils.doPrivileged(() -> deviceConnectionService.getConfigValues(user.getScopeId()));
+                deviceConnectionAuthType = (String) deviceConnectionServiceConfigValues.get("deviceConnectionAuthenticationType");
+            }
+
+            // Convert AuthRequest to LoginCredentials
+            DeviceConnectionCredentialAdapter deviceConnectionCredentialAdapter = deviceConnectionAuthHandlers.get(deviceConnectionAuthType);
+            if (deviceConnectionCredentialAdapter == null) {
+                throw new UnsupportedOperationException("No DeviceConnectionCredentialAdapter has been found for the given DeviceConnection.authenticationType: " + deviceConnectionAuthType);
+            }
+
+            LoginCredentials authenticationCredentials = deviceConnectionCredentialAdapter.mapToCredential(authRequest);
+
+            // Login
+            AccessToken accessToken = authenticationService.login(authenticationCredentials);
+
+            // Build response
+            Account account = KapuaSecurityUtils.doPrivileged(() -> accountService.find(accessToken.getScopeId()));
```

Comment:
Pure style suggestion:
I would "hide" blocks of logic behind smaller separate private methods, so that the flow is more readable;
e.g.: rows 104-118 would read more easily like this:
String deviceConnectionAuthType = extractAuthType(deviceConnection);
DeviceConnectionCredentialAdapter deviceConnectionCredentialAdapter = chooseAdapter(deviceConnectionAuthType);

Reply:
Done!